mrmustard/lab_dev/circuit_components.py

```diff
@@ -233,6 +233,34 @@ def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
 
         return CircuitComponent.from_attributes("", wires_ret, representation_ret)
 
+    def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
+        r"""
+        Contracts ``self`` and ``other`` as it would in a circuit, adding the adjoints when
+        they are missing.
+        """
+        if self.wires.ket and self.wires.bra and other.wires.ket and other.wires.bra:
+            # self has ket and bra, other has ket and bra
+            return self @ other
+
+        if self.wires.ket and self.wires.bra and (other.wires.ket or other.wires.bra):
+            # self has ket and bra, other has ket or bra
+            return self @ other @ other.adjoint
```

Comment:
Does this work in either case? I mean either if `other` has only ket or only bra?

Reply:
This is something anti-heuristic to me. I don't know how to see an object like this. 

As I understand here, this is to add the adjoint according to if one has only bra to ket but the other one has both.

Reply:
If other has ket AND bra, then line 243 returns. The return in line 247 happens if other has either ket or bra.
If you guys think that the version below is more readable, I can change it

```

        if self.wires.ket and self.wires.bra and other.wires.ket and other.wires.bra:
            # self has ket and bra, other has ket and bra
            return self @ other

        if self.wires.ket and self.wires.bra and other.wires.ket and not other.wires.bra:
            # self has ket and bra, other has ket
            return self @ other @ other.adjoint
```

Reply:
@yuan I am not sure what you mean by anti-heuristic, but line 247 is what is returned when you do ``attenuator >> dgate``

Reply:
Anti-heuristic to me means one thing only has bra no ket..

Reply:
@sylviemonet That is true, but I think we should always implement the most general case.

Reply:
This is checking for bra/ket globally, should't we only check self.output other.input?